From 62d5ab995e738f366e2f426a3a708331cba6baae Mon Sep 17 00:00:00 2001
From: donghuawei <donghuawei@eswincomputing.com>
Date: Wed, 23 Oct 2024 17:22:11 +0800
Subject: [PATCH 324/652] WIN2030-16551:feat: npu drv support device frequency

Changelogs:
	npu driver add devfreq interface.

Change-Id: I0e0b75175e9dd65e1632b3aaebeb032e6885a0f4
Signed-off-by: donghuawei <donghuawei@eswincomputing.com>
(cherry picked from commit 918be9cf0da1c2ba88cdf79ef0bbf2c5a9fb5e67)
---
 .../dts/eswin/eswin-win2030-die0-soc.dtsi     |  22 ++-
 .../dts/eswin/eswin-win2030-die1-soc.dtsi     |  14 ++
 drivers/memory/eswin/codacache/llc_spram.c    |  18 --
 drivers/soc/eswin/ai_driver/npu/dla_driver.h  |   7 +
 drivers/soc/eswin/ai_driver/npu/npu_main.c    | 159 ++++++++++++++++--
 drivers/soc/eswin/ai_driver/npu/nvdla_hw.c    |  68 +++++++-
 6 files changed, 249 insertions(+), 39 deletions(-)

diff --git a/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi b/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi
index a7e0b18764fb..49dcfe4d8eab 100644
--- a/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi
+++ b/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi
@@ -462,6 +462,18 @@ noc {
 			ranges;
 			#include "eswin-win2030-die0-noc.dtsi"
 		};
+		d0_npu_opp_table: opp-table {
+			compatible = "operating-points-v2";
+			opp@1040000000 {
+				opp-hz = /bits/ 64 <1040000000>;
+				opp-microvolt = <800000>;
+			};
+			opp@1500000000 {
+				opp-hz = /bits/ 64 <1500000000>;
+				opp-microvolt = <1050000>;
+			};
+
+		};
 
 		d0_npu: eswin-npu@51c00000 {
 			compatible = "eswin,npu";
@@ -479,11 +491,17 @@ d0_npu: eswin-npu@51c00000 {
 			clocks = <&d0_clock WIN2030_CLK_NPU_ACLK>,
 					<&d0_clock WIN2030_CLK_NPU_CFG_CLK>,
 					<&d0_clock WIN2030_CLK_NPU_CLK>,
-					<&d0_clock WIN2030_CLK_NPU_E31_CLK>;
-			clock-names = "aclk", "cfg_clk", "core_clk", "e31_core_clk";
+					<&d0_clock WIN2030_CLK_NPU_E31_CLK>,
+					<&d0_clock WIN2030_MUX_U_NPU_CORE_3MUX1_GFREE>,
+					<&d0_clock WIN2030_SPLL2_FOUT2>,
+					<&d0_clock WIN2030_SPLL1_FOUT1>;
+
+			clock-names = "aclk", "cfg_clk", "core_clk", "e31_core_clk", "mux_u_npu_core_3mux1_gfree", "fixed_rate_clk_spll2_fout2",
+					"fixed_rate_clk_spll1_fout1";
 
 			resets = <&d0_reset NPU_RST_CTRL SW_NPU_E31CORE_RSTN>;
 			reset-names = "e31_core";
+			operating-points-v2 = <&d0_npu_opp_table>;
 
 			numa-node-id = <0>;
 			firmware-name = "eic7700_die0_e31_fw";
diff --git a/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi b/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi
index 66c2a8dddd04..c1d0a7c28574 100644
--- a/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi
+++ b/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi
@@ -904,6 +904,19 @@ d1_aon_i2c1: i2c@71838000 {
 				interrupt-parent = <&plic1>;
 		};
 
+		d1_npu_opp_table: opp-table {
+			compatible = "operating-points-v2";
+			opp@1040000000 {
+				opp-hz = /bits/ 64 <1040000000>;
+				opp-microvolt = <800000>;
+			};
+			opp@1500000000 {
+				opp-hz = /bits/ 64 <1500000000>;
+				opp-microvolt = <1050000>;
+			};
+
+		};
+
 		d1_npu: eswin-npu@71c00000 {
 			compatible = "eswin,npu";
 			reg = <0x0 0x71c00000 0x0 0x400000>;
@@ -925,6 +938,7 @@ d1_npu: eswin-npu@71c00000 {
 
 			resets = <&d1_reset NPU_RST_CTRL SW_NPU_E31CORE_RSTN>;
 			reset-names = "e31_core";
+			operating-points-v2 = <&d1_npu_opp_table>;
 
 			numa-node-id = <1>;
 			firmware-name = "eic7700_die1_e31_fw";
diff --git a/drivers/memory/eswin/codacache/llc_spram.c b/drivers/memory/eswin/codacache/llc_spram.c
index 7b4301688383..8ef878c0abef 100644
--- a/drivers/memory/eswin/codacache/llc_spram.c
+++ b/drivers/memory/eswin/codacache/llc_spram.c
@@ -782,30 +782,13 @@ static int llc_clk_set_parent(struct platform_device *pdev)
 			return ret;
 		}
 	}
-
 	if (0 == spram->is_low_freq)
 	{
-		ret = regulator_set_voltage(spram->npu_regulator, NPU_1P5G_VOLTAGE, NPU_1P5G_VOLTAGE);
-		dev_dbg(dev,"name:%s,volt:%d,ret:%d\n",pdev->name,NPU_1P5G_VOLTAGE,ret);
-		if(0 != ret)
-		{
-			dev_err(dev, "set volt:%duV ret:%d\n", NPU_1P5G_VOLTAGE,ret);
-			return -EINVAL;
-		}
-		/* devm_regulator_put(spram->npu_regulator); */
-		mdelay(10);
 		ret = clk_set_parent(spram->mux_u_npu_core_3mux1_gfree,
 							 spram->fixed_rate_clk_spll1_fout1);
 	}
 	else
 	{
-		if (((NULL != spram->npu_regulator)) && (!IS_ERR(spram->npu_regulator)))
-		{
-			regulator_set_voltage(spram->npu_regulator, NPU_DEFAULT_VOLTAGE, NPU_DEFAULT_VOLTAGE);
-			dev_dbg(dev,"name:%s,volt:%d,ret:%d\n",	pdev->name,NPU_DEFAULT_VOLTAGE,ret);
-			/* devm_regulator_put(spram->npu_regulator); */
-			mdelay(10);
-		}
 		ret = clk_set_parent(spram->mux_u_npu_core_3mux1_gfree,
 							 spram->fixed_rate_clk_spll2_fout2);
 	}
@@ -815,7 +798,6 @@ static int llc_clk_set_parent(struct platform_device *pdev)
 				ret);
 		return ret;
 	}
-
 	return 0;
 }
 static int llc_clk_set_frq(struct platform_device *pdev)
diff --git a/drivers/soc/eswin/ai_driver/npu/dla_driver.h b/drivers/soc/eswin/ai_driver/npu/dla_driver.h
index b098833d2e40..bb5af332c770 100644
--- a/drivers/soc/eswin/ai_driver/npu/dla_driver.h
+++ b/drivers/soc/eswin/ai_driver/npu/dla_driver.h
@@ -67,6 +67,11 @@ struct nvdla_device {
 
 	struct reset_control *rstc_e31_core;
 	struct clk *e31_core_clk;
+	struct clk *core_clk;
+	struct clk *cfg_clk;
+	struct clk *mux_u_npu_core_3mux1_gfree;
+	struct clk *fixed_rate_clk_spll2_fout2;
+	struct clk *fixed_rate_clk_spll1_fout1;
 
 	struct clk *mbox_pclk_device;
 	struct clk *mbox_pclk;
@@ -78,6 +83,8 @@ struct nvdla_device {
 	u32 mbox_tx_lock_bit;
 	u32 mbox_tx_irq_bit;
 	spinlock_t mbox_lock;
+	unsigned long rate, volt;
+	struct mutex devfreq_lock;
 
 	uint16_t *pause_op_list;
 	struct regulator *npu_regulator;
diff --git a/drivers/soc/eswin/ai_driver/npu/npu_main.c b/drivers/soc/eswin/ai_driver/npu/npu_main.c
index 8a42a21bd79d..026a89f803b8 100644
--- a/drivers/soc/eswin/ai_driver/npu/npu_main.c
+++ b/drivers/soc/eswin/ai_driver/npu/npu_main.c
@@ -73,9 +73,17 @@
 #include "dla_buffer.h"
 #include "mailbox_regs.h"
 #include "nvdla_proc.h"
+#include <linux/eswin_npu.h>
+
+#if defined(CONFIG_PM_DEVFREQ)
+#include <linux/devfreq.h>
+#include <linux/pm_opp.h>
+#endif
+
 
 MODULE_IMPORT_NS(DMA_BUF);
 #define DRIVER_NAME "eswin_npu"
+#define NPU_CORE_CLK_HIGHEST 1500000000
 
 int64_t dla_get_time_us(void)
 {
@@ -324,6 +332,93 @@ struct nvdla_device *get_nvdla_dev(int i)
 	return static_nvdla_dev[i];
 }
 
+#if defined(CONFIG_PM_DEVFREQ)
+static int npu_devfreq_target(struct device *dev, unsigned long *freq, u32 flags)
+{
+	struct nvdla_device *nvdla_dev = dev_get_drvdata(dev);
+	struct dev_pm_opp *opp;
+	unsigned long target_volt, target_rate;
+	unsigned long rate;
+	int ret;
+
+	opp = devfreq_recommended_opp(dev, freq, flags);
+	if (IS_ERR(opp)) {
+		return PTR_ERR(opp);
+	}
+
+	target_rate = dev_pm_opp_get_freq(opp);
+	target_volt = dev_pm_opp_get_voltage(opp);
+	dev_pm_opp_put(opp);
+	if (target_rate == nvdla_dev->rate) {
+		return 0;
+	}
+	mutex_lock(&nvdla_dev->devfreq_lock);
+	ret = regulator_set_voltage(nvdla_dev->npu_regulator, target_volt, target_volt);
+	if (ret) {
+		dev_err(dev, "Cannot set voltage %lu uV\n", target_volt);
+		goto out;
+	}
+
+        if (target_volt == NPU_1P5G_VOLTAGE) {
+		ret = clk_set_parent(nvdla_dev->mux_u_npu_core_3mux1_gfree, nvdla_dev->fixed_rate_clk_spll1_fout1);
+	} else if (target_volt == NPU_DEFAULT_VOLTAGE) {
+		ret = clk_set_parent(nvdla_dev->mux_u_npu_core_3mux1_gfree, nvdla_dev->fixed_rate_clk_spll2_fout2);
+	} else {
+		dev_err(dev, "Request freq %lu is not supported.\n", *freq);
+		ret = -EINVAL;
+		goto err_parent;
+	}
+	if (ret) {
+		dev_err(dev, "Cannot set target voltage %lu parent, (%d)\n", target_rate, ret);
+		goto err_parent;
+	}
+
+        rate = clk_round_rate(nvdla_dev->core_clk, target_rate);
+        ret = clk_set_rate(nvdla_dev->core_clk, rate);
+        if (ret != 0)
+        {
+		dev_err(dev, "failed to set core_clk: %d\n", ret);
+		goto err_rate;
+
+        }
+
+	nvdla_dev->rate = clk_get_rate(nvdla_dev->core_clk);
+	if (nvdla_dev->rate != target_rate) {
+		dev_err(dev, "Got wrong frequency, Request %lu, Current %lu.\n", target_rate, nvdla_dev->rate);
+		ret = -EIO;
+		goto err_rate;
+	}
+	nvdla_dev->rate = target_rate;
+	nvdla_dev->volt = target_volt;
+	mutex_unlock(&nvdla_dev->devfreq_lock);
+	return 0;
+
+err_rate:
+	clk_set_parent(nvdla_dev->mux_u_npu_core_3mux1_gfree, nvdla_dev->fixed_rate_clk_spll1_fout1);
+err_parent:
+	regulator_set_voltage(nvdla_dev->npu_regulator, nvdla_dev->volt, nvdla_dev->volt);
+out:
+	mutex_unlock(&nvdla_dev->devfreq_lock);
+	return ret;
+}
+
+static int npu_devfreq_get_cur_freq(struct device *dev, unsigned long *freq)
+{
+	struct nvdla_device *nvdla_dev = dev_get_drvdata(dev);
+
+	*freq = nvdla_dev->rate;
+	return 0;
+}
+
+static struct devfreq_dev_profile npu_devfreq_profile = {
+	.initial_freq = NPU_CORE_CLK_HIGHEST,
+	.timer = DEVFREQ_TIMER_DELAYED,
+	.polling_ms = 1000,
+	.target = npu_devfreq_target,
+	.get_cur_freq = npu_devfreq_get_cur_freq,
+};
+#endif
+
 static int32_t  npu_probe_result = 0;
 
 static int32_t edla_probe(struct platform_device *pdev)
@@ -333,6 +428,10 @@ static int32_t edla_probe(struct platform_device *pdev)
 	struct nvdla_device *nvdla_dev;
 	struct device *dev = &pdev->dev;
 	uint32_t version;
+#if defined(CONFIG_PM_DEVFREQ)
+	struct devfreq *df = NULL;
+	struct dev_pm_opp *opp;
+#endif
 
 	dla_debug("%s enter.\n", __FUNCTION__);
 #if SMALL_PEC_MAT
@@ -346,25 +445,29 @@ static int32_t edla_probe(struct platform_device *pdev)
 	nvdla_dev = devm_kzalloc(dev, sizeof(*nvdla_dev), GFP_KERNEL);
 	if (!nvdla_dev)
 		return -ENOMEM;
-	nvdla_dev->npu_regulator = devm_regulator_get(dev, "npu");
 
+	platform_set_drvdata(pdev, nvdla_dev);
+	nvdla_dev->pdev = pdev;
+	spin_lock_init(&nvdla_dev->nvdla_lock);
+	mutex_init(&nvdla_dev->task_mutex);
+	init_waitqueue_head(&nvdla_dev->event_wq);
+	mutex_init(&nvdla_dev->devfreq_lock);
+
+	nvdla_dev->npu_regulator = devm_regulator_get(dev, "npu");
 	err = regulator_enable(nvdla_dev->npu_regulator);
 	if (err < 0)
 	{
-		dla_error("npu_regulator enble:%d err\n\r",err);
+		dla_error("npu_regulator enble error:%d\n\r", err);
 		regulator_put(nvdla_dev->npu_regulator);
 		nvdla_dev->npu_regulator = NULL;
+		return err;
 	}
 
-	platform_set_drvdata(pdev, nvdla_dev);
-	nvdla_dev->pdev = pdev;
-	spin_lock_init(&nvdla_dev->nvdla_lock);
-	mutex_init(&nvdla_dev->task_mutex);
-	init_waitqueue_head(&nvdla_dev->event_wq);
 	err = npu_dt_node_resources(nvdla_dev);
 	if (err) {
 		dla_error("error, get hw resource, ret=%d\n", err);
 		platform_set_drvdata(pdev, NULL);
+		regulator_disable(nvdla_dev->npu_regulator);
 		return -EINVAL;
 	}
 
@@ -375,6 +478,7 @@ static int32_t edla_probe(struct platform_device *pdev)
 		goto err_mem0;
 	}
 
+
 	//npu configuration space, start from 0x51c00000
 	nvdla_dev->base = devm_ioremap_resource(&pdev->dev, res);
 	if (IS_ERR(nvdla_dev->base)) {
@@ -424,6 +528,36 @@ static int32_t edla_probe(struct platform_device *pdev)
 		goto err_iomap_program;
 	}
 
+#if defined(CONFIG_PM_DEVFREQ)
+	/* Add OPP table from device tree */
+	err = devm_pm_opp_of_add_table(dev);
+	if (err) {
+		dev_err(dev, "%s, %d, Failed to add OPP table, ret = %d.\n", __func__, __LINE__, err);
+		goto err_init_reset;
+	}
+	nvdla_dev->rate = clk_get_rate(nvdla_dev->core_clk);
+	opp = devfreq_recommended_opp(dev, &nvdla_dev->rate, 0);
+	if (IS_ERR(opp)) {
+		err = PTR_ERR(opp);
+		goto err_init_reset;
+	}
+
+	nvdla_dev->rate = dev_pm_opp_get_freq(opp);
+	nvdla_dev->volt = dev_pm_opp_get_voltage(opp);
+
+	regulator_set_voltage(nvdla_dev->npu_regulator, nvdla_dev->volt, nvdla_dev->volt);
+
+	dev_pm_opp_put(opp);
+
+	df = devm_devfreq_add_device(dev, &npu_devfreq_profile, "userspace", NULL);
+	if (IS_ERR(df)) {
+		err = PTR_ERR(df);
+		dev_err(dev, "%s, %d, Failed to add devfreq device, ret=%d.\n", __func__, __LINE__, err);
+		goto err_init_reset;
+	}
+#endif
+
+
 	pm_runtime_set_autosuspend_delay(dev, 5000);
 	pm_runtime_use_autosuspend(dev);
 	pm_runtime_set_active(dev);
@@ -522,6 +656,7 @@ static int32_t edla_probe(struct platform_device *pdev)
 err_iomap_emission:
 	release_mem_region(E31_EMISSION_DTIM_BASE + nvdla_dev->numa_id * NPU_DIE_REG_OFFSET, E31_EMISSION_DTIM_SIZE);
 err_mem0:
+	regulator_disable(nvdla_dev->npu_regulator);
 	npu_put_dt_resources(nvdla_dev);
 	npu_probe_result = err;
 	return err;
@@ -565,6 +700,8 @@ static int32_t __exit edla_remove(struct platform_device *pdev)
 	ret = npu_disable_clock(nvdla_dev);
 	npu_put_dt_resources(nvdla_dev);
 	npu_remove_sysfs(pdev);
+	regulator_disable(nvdla_dev->npu_regulator);
+
 	if (nvdla_dev->pause_op_list) {
 		vfree(nvdla_dev->pause_op_list);
 		nvdla_dev->pause_op_list = NULL;
@@ -597,10 +734,10 @@ int __maybe_unused npu_runtime_resume(struct device *dev)
 		return -EIO;
 	}
 	ret = npu_enable_clock(nvdla_dev);
-    if (ret) {
-        dla_error("%s, %d, enable clock err, ret = %d.\n", __func__, __LINE__, ret);
-        return ret;
-    }
+	if (ret) {
+		dla_error("%s, %d, enable clock err, ret = %d.\n", __func__, __LINE__, ret);
+		return ret;
+	}
 	npu_tbu_power(dev, true);
 
 	return ret;
diff --git a/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c b/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
index a7d7b58cc849..a33c8efbcceb 100644
--- a/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
+++ b/drivers/soc/eswin/ai_driver/npu/nvdla_hw.c
@@ -852,7 +852,44 @@ int npu_dt_node_resources(struct nvdla_device *nvdla_dev)
 	if (IS_ERR(nvdla_dev->e31_core_clk)) {
 		ret = PTR_ERR(nvdla_dev->e31_core_clk);
 		nvdla_dev->e31_core_clk = NULL;
-		dev_err(&pdev->dev, "failed to get core_clk: %d\n", ret);
+		dev_err(&pdev->dev, "failed to get e31_core_clk: %d\n", ret);
+		return ret;
+	}
+	nvdla_dev->core_clk = devm_clk_get(&pdev->dev, "core_clk");
+	if (IS_ERR(nvdla_dev->core_clk)) {
+		ret = PTR_ERR(nvdla_dev->core_clk);
+		nvdla_dev->core_clk = NULL;
+		dev_err(&pdev->dev, "failed to get core clk: %d.\n", ret);
+		return ret;
+	}
+	nvdla_dev->cfg_clk = devm_clk_get(&pdev->dev, "cfg_clk");
+	if (IS_ERR(nvdla_dev->cfg_clk)) {
+		ret = PTR_ERR(nvdla_dev->cfg_clk);
+		nvdla_dev->cfg_clk = NULL;
+		dev_err(&pdev->dev, "failed to get cfg clk: %d.\n", ret);
+		return ret;
+	}
+	nvdla_dev->mux_u_npu_core_3mux1_gfree = devm_clk_get(&pdev->dev, "mux_u_npu_core_3mux1_gfree");
+	if (IS_ERR(nvdla_dev->mux_u_npu_core_3mux1_gfree)) {
+		ret = PTR_ERR(nvdla_dev->mux_u_npu_core_3mux1_gfree);
+		nvdla_dev->mux_u_npu_core_3mux1_gfree = NULL;
+		dev_err(&pdev->dev, "failed to get mux_u_npu_core_3mux1_gfree clk: %d.\n", ret);
+		return ret;
+	}
+
+	nvdla_dev->fixed_rate_clk_spll2_fout2 = devm_clk_get(&pdev->dev, "fixed_rate_clk_spll2_fout2");
+	if (IS_ERR(nvdla_dev->fixed_rate_clk_spll2_fout2)) {
+		ret = PTR_ERR(nvdla_dev->fixed_rate_clk_spll2_fout2);
+		nvdla_dev->fixed_rate_clk_spll2_fout2 = NULL;
+		dev_err(&pdev->dev, "failed to get fixed_rate_clk_spll2_fout2 clk: %d.\n", ret);
+		return ret;
+	}
+
+	nvdla_dev->fixed_rate_clk_spll1_fout1 = devm_clk_get(&pdev->dev, "fixed_rate_clk_spll1_fout1");
+	if (IS_ERR(nvdla_dev->fixed_rate_clk_spll1_fout1)) {
+		ret = PTR_ERR(nvdla_dev->fixed_rate_clk_spll1_fout1);
+		nvdla_dev->fixed_rate_clk_spll1_fout1 = NULL;
+		dev_err(&pdev->dev, "failed to get fixed_rate_clk_spll1_fout1 clk: %d.\n", ret);
 		return ret;
 	}
 	//nvdla_dev->rstc_e31_core = devm_reset_control_get_optional_exclusive(
@@ -1006,17 +1043,29 @@ int npu_disable_clock(struct nvdla_device *ndev)
 {
 	npu_disable_mbox_clock(ndev);
 	clk_disable_unprepare(ndev->e31_core_clk);
-	return npu_clk_gate_set(ndev->numa_id, false);
+
+	clk_disable_unprepare(ndev->core_clk);
+
+	clk_disable_unprepare(ndev->cfg_clk);
+	return 0;
 }
 
 int npu_enable_clock(struct nvdla_device *ndev)
 {
 	int ret;
-	ret = npu_clk_gate_set(ndev->numa_id, true);
-	if (ret < 0) {
-		dla_error("npu_clk_gate_set failed.\n");
-		return ret;
-	}
+
+        ret = clk_prepare_enable(ndev->cfg_clk);
+        if (ret) {
+                dla_error("failed to enable cfg_clk: %d\n", ret);
+                return ret;
+        }
+
+	ret = clk_prepare_enable(ndev->core_clk);
+        if (ret) {
+                dla_error("failed to enable core_clk: %d\n", ret);
+                goto core_clk;
+        }
+
 	ret = clk_prepare_enable(ndev->e31_core_clk);
 	if (ret < 0) {
 		dla_error("npu enable e31 core clk err.\n");
@@ -1027,11 +1076,14 @@ int npu_enable_clock(struct nvdla_device *ndev)
 		dla_error("npu enable mbox clock failed.\n");
 		goto err_mbox_clk;
 	}
+
 	return 0;
 err_mbox_clk:
 	clk_disable_unprepare(ndev->e31_core_clk);
 err_e31_clk:
-	npu_clk_gate_set(ndev->numa_id, false);
+	clk_disable_unprepare(ndev->core_clk);
+core_clk:
+	clk_disable_unprepare(ndev->cfg_clk);
 	return ret;
 }
 
-- 
2.54.0

